public/app/features/chat/Chat.tsx

```diff
@@ -0,0 +1,293 @@
+    if (!this.state.value) {
+      return;
+    }
+    await getBackendSrv().post('/api/chats/send-message/', {
+      objectId: this.props.objectId,
+      contentTypeId: this.props.contentTypeId,
+      content: this.state.value,
+    });
+    this.setState({
+      value: '',
+    });
+  };
+
+  getLiveAddr = () => {
+    return {
+      scope: LiveChannelScope.Grafana,
+      namespace: 'chat', // holds on to the last value
+      path: `${this.props.contentTypeId}/${this.props.objectId}`,
+    };
+  };
+
+  getLiveChannel = () => {
+    const live = getGrafanaLiveSrv();
+    if (!live) {
+      console.error('Grafana live not running, enable "live" feature toggle');
+      return undefined;
+    }
+
+    const addr = this.getLiveAddr();
+    if (!addr) {
+      return undefined;
+    }
+    return live.getStream<MessagePacket>(addr);
+  };
+
+  onKeyboardPress = async (e: any) => {
+    if (e.key === 'Enter' && e.shiftKey) {
+      e.preventDefault();
+      let start = e.target.selectionStart,
+        end = e.target.selectionEnd;
+      this.setState(
+        (prevState) => ({ value: prevState.value.substring(0, start) + '\n' + prevState.value.substring(end) }),
+        () => {
+          this.chatInput.selectionStart = this.chatInput.selectionEnd = start + 1;
+        }
+      );
+    } else if (e.key === 'Enter') {
+      e.preventDefault();
+    }
+  };
+
+  onKeyboardUp = async (e: any) => {
+    if (e.key === 'Enter') {
+      if (e.shiftKey) {
+        return;
+      }
+      e.preventDefault();
+      await this.sendMessage();
+    }
+  };
+
+  focusInput = () => {
+    if (!this.chatInput) {
+      return;
+    }
+    this.chatInput.focus();
+  };
+
+  scrollToBottom = () => {
+    if (!this.chatContainer) {
+      return;
+    }
+    this.chatContainer.parentNode.scrollTop = this.chatContainer.scrollHeight;
+  };
+
+  updateSubscription = () => {
+    if (this.subscription) {
+      this.subscription.unsubscribe();
+      this.subscription = undefined;
+    }
+
+    const c = this.getLiveChannel();
+    if (c) {
+      this.subscription = c.subscribe({
+        next: (msg) => {
+          if (isLiveChannelMessageEvent(msg)) {
+            const { messageCreated } = msg.message;
+            if (messageCreated) {
+              this.setState((prevState) => ({
+                messages: [...prevState.messages, messageCreated],
+              }));
+              this.scrollToBottom();
+            }
+          }
+        },
+      });
+    }
+  };
+
+  render() {
+    let messageElements;
+    if (this.state.messages.length > 0) {
+      messageElements = (
+        <div className="chat-messages">
+          {this.state.messages.map((msg) => (
+            <ChatMessage key={msg.id} message={msg} actions={this.props.actions} />
+          ))}
+        </div>
+      );
+    } else {
+      messageElements = <div className="chat-messages">No messages here yet</div>;
+    }
+
+    return (
+      <div
+        className="chat"
```

Comment:
For custom styles, instead of including in global styles file we can add styling via our `useStyles2` hook. See [docs here](https://github.com/grafana/grafana/blob/main/contribute/style-guides/styling.md). To use hook we would want to convert this to a functional component, which might be worth doing anyways for simplifying structure.

I would recommend following component structure:
`ChatManager` -> top level state management component
`Chat` or something like `ChatContainer` -> What this component currently is but just rendering chat UI / messages, and using passed down state from `ChatManager`
`ChatMessage` - just like you currently are doing, but in a separate file

Reply:
In this case, to control scroll and input focus of `ChatContainer ` from `ChatManager` we should use child refs, like:

```
this.refs.chatContainer.refs.chatInput.focus();
```

?

Or there is a better way?

Reply:
when do you want to focus input? looks like its only on mount, could it not be handled in the component? If you need to control it from the parent you could do it by passing a property "focusCounter" when it changes the component will re-focus the input